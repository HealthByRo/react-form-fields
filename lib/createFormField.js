'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = createFormField;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _snakeCase = require('lodash/snakeCase');

var _snakeCase2 = _interopRequireDefault(_snakeCase);

var _elements = require('./elements');

var _wrappers = require('./wrappers');

var _Errors = require('./Errors');

var _Errors2 = _interopRequireDefault(_Errors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function createFormField(WrappedComponent) {
  return function (_PureComponent) {
    _inherits(FormField, _PureComponent);

    function FormField(props) {
      _classCallCheck(this, FormField);

      var _this = _possibleConstructorReturn(this, (FormField.__proto__ || Object.getPrototypeOf(FormField)).call(this, props));

      if (!props.name || props.name.length < 1) {
        // TODO create file with errors, add MissingParamError
        throw new Error('name param is required for ' + WrappedComponent.name);
      }
      return _this;
    }

    _createClass(FormField, [{
      key: 'render',
      value: function render() {
        var _composeProps = composeProps(this.props),
            labelProps = _composeProps.labelProps,
            inputProps = _composeProps.inputProps,
            errorsProps = _composeProps.errorsProps;

        var showLabel = this.props.label || this.props.labelProps;

        return _react2.default.createElement(
          _wrappers.FieldWrapper,
          null,
          showLabel && _react2.default.createElement(_elements.LabelElement, labelProps),
          _react2.default.createElement(WrappedComponent, inputProps),
          _react2.default.createElement(_Errors2.default, errorsProps)
        );
      }
    }]);

    return FormField;
  }(_react.PureComponent);
}

function composeProps(props) {
  var name = props.name,
      placeholder = props.placeholder,
      label = props.label,
      errors = props.errors,
      inputProps = _objectWithoutProperties(props, ['name', 'placeholder', 'label', 'errors']);

  inputProps.id = inputProps.id || generateInputId(inputProps);
  inputProps.name = name;

  var labelProps = {
    htmlFor: inputProps.id,
    text: label
  };

  var errorsProps = {
    errors: errors
  };

  // don't set placeholder when undefined
  // to avoid placeholder={undefined} in rendered html
  if (placeholder) {
    inputProps.placeholder = placeholder;
  }

  return {
    labelProps: labelProps,
    inputProps: inputProps,
    errorsProps: errorsProps
  };
}

function generateInputId(props) {
  return 'id_' + (0, _snakeCase2.default)(props.name);
}