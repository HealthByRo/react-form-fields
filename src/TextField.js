/* @flow */

import React from 'react';
import BaseField from './BaseField';
import { TextFieldElement } from './elements';
import type { FieldProps } from './types';

export default class TextField extends BaseField {
  throwMissingNameError() {
    // TODO create file with errors, add MissingParamError
    throw new Error('name param is required for TextField');
  }

  renderFieldElement(props: FieldProps) {
    return (
      <TextFieldElement {...props} />
    );
  }

  render() {
    return super.renderWithField(this.renderFieldElement);
  }
}
