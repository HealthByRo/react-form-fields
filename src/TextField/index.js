/* @flow */

import React from 'react';
import { composeInputProps } from '../helpers';
import type { InputProps, TextFieldElementProps } from '../types';

const TextField = (props: TextFieldElementProps) => {
  const inputProps: InputProps = composeInputProps('text', props);

  return (
    <input type="text" {...inputProps} />
  );
};
// TODO figure out a better way of using defaults in stateless component
TextField.defaultProps = {
  className: 'textfield',
};

export default TextField;
