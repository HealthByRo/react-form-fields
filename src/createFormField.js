/* @flow */

import React, { PureComponent } from 'react';
import snakeCase from 'lodash/snakeCase';
import { LabelElement } from './elements';
import { FieldWrapper } from './wrappers';
import Errors from './Errors';
import type { FieldProps, LabelProps, InputProps, ErrorsProps, Component } from './types';

export default function createFormField(WrappedComponent: Component<*, FieldProps, *>) {
  return class FormField extends PureComponent {
    constructor(props: FieldProps) {
      super(props);

      if (!props.name || props.name.length < 1) {
        // TODO create file with errors, add MissingParamError
        throw new Error(`name param is required for ${WrappedComponent.name}`);
      }
    }

    props: FieldProps;

    render() {
      const {
        labelProps,
        inputProps,
        errorsProps,
      } = composeProps(this.props);
      const showLabel = this.props.label || this.props.labelProps;

      return (
        <FieldWrapper>
          {showLabel &&
            <LabelElement {...labelProps} />
          }
          <WrappedComponent {...inputProps} />
          <Errors {...errorsProps} />
        </FieldWrapper>
      );
    }
  };
}

function composeProps(props: FieldProps): Object {
  const {
    name,
    placeholder,
    label,
  } = props;
  const inputId = props.inputId || generateInputId(props);
  const labelProps: LabelProps = {
    htmlFor: inputId,
    text: label,
  };
  const inputProps: InputProps = {
    name,
    id: inputId,
  };
  const errorsProps: ErrorsProps = {
    errors: props.errors,
  };

  // don't set placeholder when undefined
  // to avoid placeholder={undefined} in rendered html
  if (placeholder) {
    inputProps.placeholder = placeholder;
  }

  return {
    labelProps,
    inputProps,
    errorsProps,
  };
}

function generateInputId(props: FieldProps): string {
  return `id_${snakeCase(props.name)}`;
}
