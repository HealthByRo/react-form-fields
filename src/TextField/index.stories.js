import React from 'react';
import { storiesOf } from '@kadira/storybook';
import TextField from './';

storiesOf('Fields: TextField', module)
  .add('input', () => {
    const props = {
      name: 'TEXT_FIELD_NAME',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder', () => {
    const props = {
      name: 'TEXT_FIELD_NAME',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder + error', () => {
    const props = {
      errors: ['Sample error for field'],
      name: 'TEXT_FIELD_NAME',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder + multiple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      name: 'TEXT_FIELD_NAME',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label', () => {
    const props = {
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + error', () => {
    const props = {
      errors: ['Sample error for field'],
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + multiple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder', () => {
    const props = {
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder + error', () => {
    const props = {
      errors: ['Sample error for field'],
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder + mutliple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      name: 'TEXT_FIELD_NAME',
      label: 'Sample label',
      placeholder: 'Sample text field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  });
