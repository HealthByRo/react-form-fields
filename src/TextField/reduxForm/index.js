/* @flow */

import TextField from '../';
import createReduxFormField from '../../createReduxFormField';

export default createReduxFormField(TextField);
