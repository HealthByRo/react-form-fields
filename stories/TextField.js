import React from 'react';
import { storiesOf } from '@kadira/storybook';
import TextField from '../src/TextField';

storiesOf('Fields.TextField', module)
  .add('input', () => {
    const props = {
      inputId: 'ID-I',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder', () => {
    const props = {
      inputId: 'ID-IP',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder + error', () => {
    const props = {
      errors: ['Sample error for field'],
      inputId: 'ID-IPE',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + placeholder + multiple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      inputId: 'ID-IPES',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label', () => {
    const props = {
      inputId: 'ID-IL',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + error', () => {
    const props = {
      errors: ['Sample error for field'],
      inputId: 'ID-ILE',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + multiple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      inputId: 'ID-ILES',
      label: 'Sample label',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder', () => {
    const props = {
      inputId: 'ID-ILP',
      label: 'Sample label',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder + error', () => {
    const props = {
      errors: ['Sample error for field'],
      inputId: 'ID-ILPE',
      label: 'Sample label',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  })
  .add('input + label + placeholder + mutliple errors', () => {
    const props = {
      errors: [
        'First error for field',
        'Second error for field',
      ],
      inputId: 'ID-ILPES',
      label: 'Sample label',
      placeholder: 'Sample input field with placeholder',
    };

    return (
      <TextField {...props} />
    );
  });
