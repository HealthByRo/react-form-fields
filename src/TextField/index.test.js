/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import TextField from './';
import createFormField from '../createFormField';
import type { FieldProps } from '../types';

describe('<TextField />', () => {
  let wrapper;
  const defaultProps = {
    name: 'INPUT_NAME',
    inputId: 'INPUT_ID',
  };
  const shallowWithProps = (props: FieldProps) => {
    wrapper = shallow(
      <TextField {...props} />
    );
  };

  describe('html element', () => {
    describe('just name and inputId', () => {
      beforeEach(() => {
        shallowWithProps(defaultProps);
      });

      it('should be input', () => {
        expect(wrapper.name()).toBe('input');
      });

      it('should have [type=text]', () => {
        expect(wrapper.props().type).toEqual('text');
      });

      it('should have [name=INPUT_NAME]', () => {
        expect(wrapper.props().name).toEqual('INPUT_NAME');
      });

      it('should have default .textfield class', () => {
        expect(wrapper.hasClass('textfield')).toBe(true);
      });

      it('should not have [placeholder]', () => {
        expect(wrapper.props.placeholder).not.toBeDefined();
      });

      it('should not have [placeholder]', () => {
        expect(wrapper.props.placeholder).not.toBeDefined();
      });
    });
  });

  describe('more params', () => {
    it('should use className from props', () => {
      shallowWithProps({
        ...defaultProps,
        className: 'CSS_CLASS',
      });

      expect(wrapper.hasClass('CSS_CLASS')).toBe(true);
      expect(wrapper.hasClass('textfield')).toBe(false);
    });

    it('should use placeholder from props', () => {
      shallowWithProps({
        ...defaultProps,
        placeholder: 'PLACEHOLDER',
      });

      expect(wrapper.props().placeholder).toEqual('PLACEHOLDER');
    });

    it('should use id from props.inputId', () => {
      shallowWithProps({
        ...defaultProps,
        inputId: 'INPUT_ID',
      });

      expect(wrapper.props().id).toEqual('INPUT_ID');
    });
  });
});

describe('createFormField(TextField)', () => {
  let tree;
  const FormTextField = createFormField(TextField);
  const placeholder = 'INPUT_PLACEHOLDER';
  const label = 'INPUT_LABEL';
  const renderWithProps = (props: FieldProps) => {
    tree = renderer.create(
      <FormTextField {...props} />
    ).toJSON();
  };
  const getPropsWithName = (props?: Object, name?: string = 'INPUT_NAME'): FieldProps => {
    const propsWithName = { ...props };

    if (!propsWithName.name) {
      propsWithName.name = name;
    }

    return propsWithName;
  };

  describe('<TextField />', () => {
    it('should throw error if no name passed', () => {
      // $FlowExpectedError
      expect(() => renderWithProps()).toThrowError('name param is required for TextField');
    });
  });

  describe('<TextField { name } />', () => {
    it('should NOT throw any errors if name is passed', () => {
      expect(() => <TextField name="TEST_NAME" />).not.toThrowError();
    });

    it('should render only wrapped input field with defaults', () => {
      renderWithProps(getPropsWithName());

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder } />', () => {
    it('should render input with placeholder', () => {
      renderWithProps(getPropsWithName({ placeholder }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder, errors: [SINGLE_ERROR] } />', () => {
    it('should render input with placeholder and single error', () => {
      renderWithProps(getPropsWithName({ placeholder, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder, errors: [FIRST_ERROR, SECOND_ERROR] } />', () => {
    it('should render input with placeholder and two errors', () => {
      renderWithProps(getPropsWithName({ placeholder, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, label } />', () => {
    it('should render input and label', () => {
      renderWithProps(getPropsWithName({ label }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, label, errors: [SINGLE_ERROR] } />', () => {
    it('should render input and label and single error', () => {
      renderWithProps(getPropsWithName({ label, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, label, errors: [FIRST_ERROR, SECOND_ERROR] } />', () => {
    it('should render input and label and two errors', () => {
      renderWithProps(getPropsWithName({ label, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder, label } />', () => {
    it('should render input with placeholder and label', () => {
      renderWithProps(getPropsWithName({ label, placeholder }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder, label, errors: [SINGLE_ERROR] } />', () => {
    it('should render input with placeholder and label and single error', () => {
      renderWithProps(getPropsWithName({ label, placeholder, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TextField { name, placeholder, label, errors: [FIRST_ERROR, SECOND_ERROR] } />', () => {
    it('should render input with placeholder and label and two errors', () => {
      renderWithProps(getPropsWithName({ label, placeholder, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });
});
