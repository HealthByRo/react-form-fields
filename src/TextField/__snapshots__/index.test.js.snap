// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFormField(TextField) <TextField { name } /> should render only wrapped input field with defaults 1`] = `
<div
  className="form-item"
>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    type="text"
  />
</div>
`;

exports[`createFormField(TextField) <TextField { name, label } /> should render input and label 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    type="text"
  />
</div>
`;

exports[`createFormField(TextField) <TextField { name, label, errors: [FIRST_ERROR, SECOND_ERROR] } /> should render input and label and two errors 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TextField) <TextField { name, label, errors: [SINGLE_ERROR] } /> should render input and label and single error 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder } /> should render input with placeholder 1`] = `
<div
  className="form-item"
>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder, errors: [FIRST_ERROR, SECOND_ERROR] } /> should render input with placeholder and two errors 1`] = `
<div
  className="form-item"
>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder, errors: [SINGLE_ERROR] } /> should render input with placeholder and single error 1`] = `
<div
  className="form-item"
>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder, label } /> should render input with placeholder and label 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder, label, errors: [FIRST_ERROR, SECOND_ERROR] } /> should render input with placeholder and label and two errors 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TextField) <TextField { name, placeholder, label, errors: [SINGLE_ERROR] } /> should render input with placeholder and label and single error 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="TF-INPUT_NAME"
  >
    INPUT_LABEL
  </label>
  <input
    className="textfield"
    id="TF-INPUT_NAME"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;
